### PR TITLE
Mock array big tweaks

### PR DESCRIPTION
**Pull request by oharboe**
more readable code and a small fix so that the spacing around the array is the same horizontally and vertically.

### CODE REVIEW

**Review by maliberty — APPROVED** [review]

**Comment by maliberty** [comment]
Fails in placement

**Comment by oharboe** [comment]
> Fails in placement

Hmm... Why?

The floorplan looks good.. I added a bit of more space here, but it times out: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1096

Error:

```
[INFO GPL-0020] StdInstsArea: 102060
[INFO GPL-0021] MacroInstsArea: 0
[ERROR GPL-0301] Utilization exceeds 100%.
Error: global_place_skip_io.tcl, 30 GPL-0301
Command exited with non-zero status 1
Elapsed time: 0:01.53[h:]min:sec. CPU time: user 1.44 sys 0.08 (99%). Peak memory: 196900KB.
make: *** [Makefile:513: results/asap7/mock-array-big/base/3_1_place_gp_skip_io.odb] Error 1
```



image

**Comment by maliberty** [comment]
Its a bit hard to tell from the image but it looks like everything is covered in placement blockage.

**Comment by oharboe** [comment]
image

File: flow/designs/asap7/mock-array-big/macro-placement.tcl [hunk 1]
```diff
@@ -9,8 +9,8 @@ for {set row 0} {$row < $rows} {incr row} {
   for {set col 0} {$col < $cols} {incr col} {
     set inst [$block findInst [format "ces_%d_%d" $row $col]]
 
-    set x [expr int(($margin_x + ($ce_pitch_x / 2) + ($ce_pitch_x * $col)) * $units)]
-    set y [expr int(($margin_y + ($ce_pitch_y / 2) + ($ce_pitch_y * $row)) * $units)]
+    set x [expr int(int(($margin_x + ($ce_pitch_x / 2) + ($ce_pitch_x * $col)) / $placement_grid_x) * $placement_grid_x * $units)]
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/macro-placement.tcl:12]: @maliberty I'm struggling with figuring out what grid to align the macro placement to. So far I have been trawling files and guessing. Ideally there would be a tcl interface that gave me the number to use here...

**Comment by oharboe** [comment]
After update(stopped it after 5 itreations so it still has some DRC errors, but seems to be converging).

The pins seem to be landing nicely into the macro:

image

image

**Comment by oharboe** [comment]
After 64 iterations, there are still DRC errors:

image

zooming in on a random one. They seem very regular.... why not elsewhere in the array?

image

Also, the element macro, has DRC errors:

```
make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk
make DESIGN_CONFIG=designs/asap7/mock-array-big/Element/config.mk gui_final
```

These errors are also around the pins. Perhaps fastest to look at the Element macro DRC errors, if they have a common root cause...

image

**Comment by oharboe** [comment]
@maliberty Can you have a look at the DRC errors?

**Comment by maliberty** [comment]
The first violation I looked at is:
image

clearly the power stripe is too close to the pin and blocking it.  The pdn.tcl:
```
# The halo around the macro prevents pdn from blocking pin access
define_pdn_grid -macro -cells Element -halo "1.0 1.0 1.0 1.0" -voltage_domains {CORE} -name ElementGrid
```
is not a large enough halo.  I tried changing the halo to 3.0 but I see:
image

@gadfort any idea why only VSS seems to be obeying the halo?

**Comment by gadfort** [comment]
@maliberty I can try to take a look later today, but I don't see an obvious reason why VSS is acting differently

**Comment by gadfort** [comment]
@oharboe how do I get to the error you are seeing. I only see the instructions for the "Element" but nothing about how I might trigger the big design.
Please do not forget to include that information when requesting help since it makes it harder to provide assistance (especially when we do so in our free time).

**Comment by oharboe** [comment]
> @oharboe how do I get to the error you are seeing. I only see the instructions for the "Element" but nothing about how I might trigger the big design. Please do not forget to include that information when requesting help since it makes it harder to provide assistance (especially when we do so in our free time).

@maliberty and I have been collaborating on whitling down a number of issues with mock-array-big in master. 

I asked @maliberty why we had DRC errors, because that's beyond my skillset currently. The instructions I gave was for Element, because it should not have any DRC errors either.

Matt went beyond studying DRC in Element apparently. Therefore I must refer to him w.r.t. how to reproduce the exact problem he saw with VSS and how to find it.

Note! @maliberty also made some local changes on top of this PR, but I tried go create those halo fixes in another commit since his comments on this PR.

That said, to build the whole mock-array-big(make issue does not work for designs with macros when I tried it), run:


```
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk gui_final
```

The Element macro will be built as part of the build above. Note that a `make clean_all` does not delete macros, so the safe thing is to delete the entire objects/ results/ and logs/ folders to rebuild everything from scratch.

**Comment by oharboe** [comment]
@gadfort Ah. You can navigate to and download artifacts for this builds... No need to build locally. 

Download: 
https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/public_tests_all-pr/job/PR-1097-head/4/artifact/flow/final_report_mock-array-big_asap7_base_2023-05-29_02-38.tar.gz

Extract and view results:

```
tar --strip-components=1 -xzvf ~/Downloads/final_report_mock-array-big_asap7_base_2023-05-29_02-38.tar.gz
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk gui_final
```

Load DRC report:

image

zoom in:

image

**Comment by gadfort** [comment]
> @gadfort Ah. You can navigate to and download artifacts for this builds... No need to build locally.
> 
> Download: https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/public_tests_all-pr/job/PR-1097-head/4/artifact/flow/final_report_mock-array-big_asap7_base_2023-05-29_02-38.tar.gz
> 
> Extract and view results:
> 
> ```
> tar --strip-components=1 -xzvf ~/Downloads/final_report_mock-array-big_asap7_base_2023-05-29_02-38.tar.gz
> make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk gui_final
> ```
> 
> Load DRC report:
> 
> image
> 
> zoom in:
> 
> image

@oharboe thanks. I'm sorry for getting frustrated earlier. I'll try to take a look at the issue @maliberty highlighted later this week

**Comment by oharboe** [comment]
@gadfort Hi Peter, a gentle a kind reminder to have a look at this if you have the time. :-)

**Comment by gadfort** [comment]
@maliberty would it be possible to open an issue with the powerstrap problem with a runnable testcase?

**Comment by yupferris** [comment]
Rebased on master (removes the need for d3264fd4ef4e7d439343c920177361f9a73f72a4); the problem still remains. I'm looking into getting a runnable test case.

**Comment by yupferris** [comment]
Following the guide in the issue form, with `ISSUE_TAG=mock-array-big-pdn-issue` and `make DESIGN_CONFIG=./designs/asap7/mock-array-big/config.mk pdn_issue`, [this archive](https://www.dropbox.com/s/fzm7qw5i6kjgfcv/pdn_mock-array-big-pdn-issue.tar.gz?dl=0) was produced (NOTE: this is a dropbox link, as the file was too large (181MB) to upload to GitHub).

I'm a bit skeptical that `pdn` was the right stage to isolate, but I'm not sure which would have been better in this case, if any.

**Comment by gadfort** [comment]
@yupferris thanks, I'll take a look this afternoon

**Comment by gadfort** [comment]
@yupferris  Please try this: https://github.com/The-OpenROAD-Project/OpenROAD/pull/3432

**Comment by yupferris** [comment]
> @yupferris Please try this: https://github.com/The-OpenROAD-Project/OpenROAD/pull/3432

Cheers! I've rebased on master and added [3a0a1ef](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1097/commits/3a0a1efd1d8d7891de1c4961487eaf6288adf7df) to upgrade to latest OpenROAD, which contains these fixes. It seems indeed that the PDN issue above is resolved.

However, there seems to be a new DRC error:
image
image

It looks as though there's a net/via placed too close to the edge of a macro instance:

image

What might be causing this?

**Comment by yupferris** [comment]
I've created another test case, this time with `ISSUE_TAG=mock-array-big-drc-error` and `make DESIGN_CONFIG=./designs/asap7/mock-array-big/config.mk detail_route_issue`, producing [this archive](https://www.dropbox.com/s/cemqcjgfixdwmu6/detail_route_mock-array-big-drc-error.tar.gz?dl=0) (NOTE: another DropBox link due to large file size).

As before I'm a bit skeptical about choosing `detail_route` as the stage to produce the case with, but it seems reasonable enough.

**Comment by oharboe** [comment]
@yupferris Try increasing number of detailed route iterations to 15 https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/aa851c044acc79024e33d73c843113b3d2ca4d81/flow/designs/asap7/mock-array-big/config.mk#L37


If detailed route succeeds w 15 iterations we are good. Though, there might be a separate concern and github issue w.r.t. detailed route taking too many iterations due to a bug somewhere. This bug should be investigated in a separate github issue by looking at DRC failures after, say, 7 iterations. In master, detailed routing succeeds after 4 iterations as I recall.

**Comment by yupferris** [comment]
> @yupferris Try increasing number of detailed route iterations to 15

Right, I didn't mention it, but I did do that before reporting, and it didn't help.

**Comment by oharboe** [comment]
Something odd is going on. I updated the branch, but the web page didn't update. I will try to close and reopend the PR.

**Comment by oharboe** [comment]
> Something odd is going on. I updated the branch, but the web page didn't update. I will try to close and reopend the PR.

That fixed it... Odd...

**Comment by oharboe** [comment]
@maliberty Could you have a look at these two DRC errors and advise?

To view the DRC errors:

1. download https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/public_tests_all-pr/job/PR-1097-head/9/artifact/flow/final_report_mock-array-big_asap7_base_2023-06-07_19-06.tar.gz
2. `tar --strip-components=1 -xzvf final_report_mock-array-big_asap7_base_2023-06-07_19-06.tar.gz`
3. `make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk gui_final` and load `5_route_drc.rpt`

There are two DRC errors that look the same:

image

image


zooming in on the same area, but on a different element without DRC errors:

image

**Comment by maliberty** [comment]
@osamahammad21 any thoughts on these drcs?

**Comment by osamahammad21** [comment]
I am looking at the DRCs and I can say the following:

- The DRCs are real. The spacing is smaller than the allowed spacing on M2.
- The design converges at iteration 17
- In general the design converges when we increase the fixed shape cost
I am still debugging why it can't solve the violation  at the early stages

**Comment by oharboe** [comment]
@osamahammad21 @maliberty Upgraded pull request to have latest master from OpenROAD, 0 DRCs after 4 iterations locally. Trying server build.

**Comment by oharboe** [comment]
@maliberty If this goes green, then it's ready to merge for my part:

```
make DESIGN_CONFIG=designs/asap7/mock-array-big/config.mk metadata
[deleted]
All metadata rules passed (22 rules)
```

**Comment by oharboe** [comment]
@maliberty Green! There are failures unrelated to mock-array-big.

image

**Review by maliberty — APPROVED** [review]

**Pull request merged**